includes/REST_API/Stories_Users_Controller.php

```diff
@@ -76,7 +78,83 @@ public static function get_registration_action() {
 	 *
 	 * @return int Registration action priority to use.
 	 */
-	public static function get_registration_action_priority() {
+	public static function get_registration_action_priority(): int {
 		return 100;
 	}
+
+	/**
+	 * Checks if a given request has access to read a user.
+	 *
+	 * Same as the parent function but with using a cached version of {@see count_user_posts()}.
+	 *
+	 * @see WP_REST_Users_Controller::get_item_permissions_check()
+	 *
+	 * @since 1.10.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access for the item, otherwise WP_Error object.
+	 */
+	public function get_item_permissions_check( $request ) {
+		$user = $this->get_user( $request['id'] );
+		if ( is_wp_error( $user ) ) {
+			return $user;
+		}
+
+		$types = get_post_types( [ 'show_in_rest' => true ], 'names' );
+
+		if ( get_current_user_id() === $user->ID ) {
+			return true;
+		}
+
+		if ( 'edit' === $request['context'] && ! current_user_can( 'list_users' ) ) {
+			return new WP_Error(
+				'rest_user_cannot_view',
+				__( 'Sorry, you are not allowed to list users.', 'web-stories' ),
+				[ 'status' => rest_authorization_required_code() ]
+			);
+		}
+
+		if ( ! $this->count_user_posts( $user->ID, $types ) && ! current_user_can( 'edit_user', $user->ID ) && ! current_user_can( 'list_users' ) ) {
+			return new WP_Error(
+				'rest_user_cannot_view',
+				__( 'Sorry, you are not allowed to list users.', 'web-stories' ),
+				[ 'status' => rest_authorization_required_code() ]
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Number of posts user has written.
+	 *
+	 * Wraps {@see count_user_posts()} results in a cache.
+	 *
+	 * @since 1.10.0
+	 * @link https://core.trac.wordpress.org/ticket/39242
+	 *
+	 * @SuppressWarnings(PHPMD.BooleanArgumentFlag)
+	 *
+	 * @param int          $userid      User ID.
+	 * @param array|string $post_type   Optional. Single post type or array of post types to count the number of posts for. Default 'post'.
+	 * @param bool         $public_only Optional. Whether to only return counts for public posts. Default false.
+	 * @return string Number of posts the user has written in this post type.
+	 */
+	protected function count_user_posts( $userid, $post_type = 'post', $public_only = false ): string {
+		$post_type = implode( '|', (array) $post_type );
+		$cache_key = "count_user_{$post_type}_{$userid}";
+		if ( $public_only ) {
+			$cache_group = 'user_posts_count_public';
+		} else {
+			$cache_group = 'user_posts_count';
+		}
+		$count = wp_cache_get( $cache_key, $cache_group );
+		if ( false === $count ) {
+			// phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.count_user_posts_count_user_posts
+			$count = count_user_posts( $userid, $post_type, $public_only );
+			wp_cache_add( $cache_key, $count, $cache_group );
+		}
+
+		return $count;
+	}
```

Comment:
How is this cache cleaned?

Reply:
That is a great question 😅 

For our use case (saving a few duplicate queries per request) a simple TTL of 1min or so would already suffice.

Alternatively, invalidate whenever a post is deleted or updated (status or visibility changes). But that's quite overkill for our use case.

Reply:
I think that do clear caches very simply. 

```php
add_action( 'clean_post_cache', function( $id, $post ){
    $cache_key = "count_user_{$post->post_type}_{$post->author_id}";
    wp_cache_delete( $cache_key, 'user_posts_count_public' );
    wp_cache_delete( $cache_key, 'user_posts_count' );
    $types = get_post_types( [ 'show_in_rest' => true ], 'names' );
    if( in_array( $post->post_type, $types, true ){
          $post_type = implode( '|', (array) $types );
          $cache_key = "count_user_{$post_type}_{$post->author_id}";
          wp_cache_delete( $cache_key, 'user_posts_count_public' );
          wp_cache_delete( $cache_key, 'user_posts_count' );
    }
});
```

```php
add_action( 'clean_user_cache', function( $id ){
     $types = get_post_types( [ 'show_in_rest' => true ], 'names' );
    $post_type = implode( '|', (array) $types );
    $cache_key = "count_user_{$post_type}_{$id}";
    wp_cache_delete( $cache_key, 'user_posts_count_public' );
    wp_cache_delete( $cache_key, 'user_posts_count' );
});

Reply:
I also think that this method should live in it's own class in the user namespace.

Reply:
Seems like you could easily contribute that to core! :)

I think here the cache group could be simplified because we only ever call it with `$public_only === false` (could even remove that parameter).
Otherwise this won't account for cases where one calls `count_user_posts()` with an arbitrary set of post types.

For extra safety, the result of `get_post_types()` could be sorted.

Reply:
Is there a way we can implement this without making it overly complicated? What would it look like in a custom class, would it be a static method?

Reply:
Here is a bash at trying to make a simple cache invalidated validation. 

https://github.com/google/web-stories-wp/commit/9bed7211b096f335f642885af65f4a37b1a0c39a

I think all we care about if a user has existing stories. 

Let me see what you think.

Reply:
Hmm probably. Let's give it a try 👍 

Feel free to push that commit to this PR directly and we can iterate from there.